src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs

```diff
@@ -0,0 +1,283 @@
+        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="fileOffset" /> is negative.</exception>
+        /// <exception cref="T:System.UnauthorizedAccessException"><paramref name="handle" /> was not opened for reading.</exception>
+        /// <exception cref="T:System.IO.IOException">An I/O error occurred.</exception>
+        /// <remarks>Position of the file is not advanced.</remarks>
+        public static long Read(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset)
+        {
+            ValidateInput(handle, fileOffset, mustBeSync: OperatingSystem.IsWindows());
+            ValidateBuffers(buffers);
+
+            return ReadScatterAtOffset(handle, buffers, fileOffset);
+        }
+
+        /// <summary>
+        /// Reads a sequence of bytes from given file at given offset.
+        /// </summary>
+        /// <param name="handle">The file handle.</param>
+        /// <param name="buffer">A region of memory. When this method returns, the contents of this region are replaced by the bytes read from the file.</param>
+        /// <param name="fileOffset">The file position to read from.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="P:System.Threading.CancellationToken.None" />.</param>
+        /// <returns>The total number of bytes read into the buffer. This can be less than the number of bytes allocated in the buffer if that many bytes are not currently available, or zero (0) if the end of the file has been reached.</returns>
+        /// <exception cref="T:System.ArgumentNullException"><paramref name="handle" /> is <see langword="null" />.</exception>
+        /// <exception cref="T:System.ArgumentException"><paramref name="handle" /> is invalid.</exception>
+        /// <exception cref="T:System.ObjectDisposedException">The file is closed.</exception>
+        /// <exception cref="T:System.NotSupportedException">The file does not support seeking (pipe or socket).</exception>
+        /// <exception cref="T:System.ArgumentException"><paramref name="handle" /> was not opened for async IO.</exception>
+        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="fileOffset" /> is negative.</exception>
+        /// <exception cref="T:System.UnauthorizedAccessException"><paramref name="handle" /> was not opened for reading.</exception>
+        /// <exception cref="T:System.IO.IOException">An I/O error occurred.</exception>
+        /// <remarks>Position of the file is not advanced.</remarks>
+        public static ValueTask<int> ReadAsync(SafeFileHandle handle, Memory<byte> buffer, long fileOffset, CancellationToken cancellationToken = default)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ValueTask.FromCanceled<int>(cancellationToken);
+            }
```

Comment:
For all of these we should do the cancellation check only after we've validated the inputs.

Reply:
Should we fix it in other places as well? Example: https://github.com/dotnet/runtime/blob/e492dc7e84df190636f3ab7812930b6cdd1b7252/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L331-L356

Reply:
In that example the buffer argument validation is correctly happening before the cancellation check.  The disposed check is a gray area and we don't need to change it.

Reply:
> The disposed check is a gray area and we don't need to change it.

I see, thanks